File: factory/PaymentProcessor.java
package factory;

public class PaymentProcessor {

    private final PaymentMethodFactory paymentMethodFactory;

    public PaymentProcessor(PaymentMethodFactory paymentMethodFactory) {
        this.paymentMethodFactory = paymentMethodFactory;
    }

    public void processPayment(CardType cardType, double amount, String cardHolder, String cardNumber,
                               String cvv, String expiryDate) {
        // logic of create the real card + processing
        PaymentMethod paymentMethod = paymentMethodFactory.createPaymentMethod(cardType,
                cardHolder, cardNumber, cvv, expiryDate);

        paymentMethod.authorizePayment();
        paymentMethod.startMoneyTransfer();
        paymentMethod.calculatePaymentFees(amount);
    }
}

File: typescript/factory/payment-processor.ts
import { CardType } from "./card-type";
import { PaymentMethod, PaymentMethodFactory } from "./payment-method-factory";

export class PaymentProcessor {
    private readonly paymentMethodFactory: PaymentMethodFactory;
  
    constructor(paymentMethodFactory: PaymentMethodFactory) {
      this.paymentMethodFactory = paymentMethodFactory;
    }
  
    processPayment(
      cardType: CardType,
      amount: number,
      cardHolder: string,
      cardNumber: string,
      cvv: string,
      expiryDate: string
    ): void {
      // Logic to create the real card + processing
      const paymentMethod: PaymentMethod =
        this.paymentMethodFactory.createPaymentMethod(
          cardType,
          cardHolder,
          cardNumber,
          cvv,
          expiryDate
        );
  
      paymentMethod.authorizePayment();
      paymentMethod.startMoneyTransfer();
      paymentMethod.calculatePaymentFees(amount);
    }
  }
  